File: python/hangover.py
table = {0: 0.0}

def f(n):
	result = 0
	length = 0.0
	i = 2
	while length < n:
		try:
			length = table[i-1]
		except (KeyError):
			length = 0.0
			for j in range (2, i+1):
				length += 1.0/j
			table[i-1] = length
		result += 1
		i += 1
	return result

while True:
	n = float (raw_input())
	if not n:
		break
	else:
		print f(n), "card(s)"

File: cplusplus/hangover.cpp
#include <iostream>
#include <map>

using namespace std;

map <int, float> table;


int f (float n)
{
	int result = 1;
	float length = 0;
	for (int i = 2; length < n; ++i, ++result)
	{
		if (table[i-1])
		{
			length = table[i-1];
		}
		else
		{
			length = 0;
			for (int j = 2; j <= i; ++j)
			{
				length += 1.0/j;
			}
		}
		table[i-1] = length;
	}
	return result-1;
}


int main()
{
	table[0] = 0;
	while (true)
	{
		float n;
		cin >> n;
		if (n)
		{
			cout << f(n) << " card(s)" << endl;
			cin.get();
		}
		else
			break;
	}
}
